### PR TITLE
Made the traffic_weight block in the ingress block of the azurerm_container_app resource optional 

### LINKED ISSUE #20537: In azurerm_container_app the traffic_weight block in the ingress block should not be required
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

v1.3.7

### AzureRM Provider Version

v3.43.0

### Affected Resource(s)/Data Source(s)

azurerm_container_app 

### Terraform Configuration Files

```hcl
variable "container_apps" {
  description = "Specifies the container apps in the managed environment."
  type = list(object({
    name                           = string
    revision_mode                  = optional(string)
    ingress                        = optional(object({
      allow_insecure_connections   = optional(bool)
      external_enabled             = optional(bool)
      target_port                  = optional(number)
      transport                    = optional(string)
      traffic_weight               = optional(list(object({
        label                      = optional(string)
        latest_revision            = optional(bool)
        revision_suffix            = optional(string)
        percentage                 = optional(number)
      })))
    }))
    dapr                           = optional(object({
      app_id                       = optional(string)
      app_port                     = optional(number)
      app_protocol                 = optional(string)
    }))
    secrets                        = optional(list(object({
      name                         = string
      value                        = string
    })))
    template                       = object({
      containers                   = list(object({
        name                       = string
        image                      = string
        args                       = optional(list(string))
        command                    = optional(list(string))
        cpu                        = optional(number)
        memory                     = optional(string)
        env                        = optional(list(object({
          name                     = string
          secret_name              = optional(string)
          value                    = optional(string)
        })))
      }))
      min_replicas                 = optional(number)
      max_replicas                 = optional(number)
      revision_suffix              = optional(string)
      volume                       = optional(list(object({
        name                       = string
        storage_name               = optional(string)
        storage_type               = optional(string)
      })))
    })
  }))
  default                          = [{
    name                           = "nodeapp"
    revision_mode                  = "Single"
    ingress                        = {
      external_enabled             = false
      target_port                  = 3000
      transport                    = "http"
      traffic_weight               = [{
        label                      = "blue"
        latest_revision            = true
        revision_suffix            = "blue"
        percentage                 = 100
      }]
    }
    dapr                           = {
      app_id                       = "nodeapp"
      app_port                     = 3000
      app_protocol                 = "http"
    }
    template                       = {
      containers                   = [{
        name                       = "hello-k8s-node"
        image                      = "dapriosamples/hello-k8s-node:latest"
        cpu                        = 0.5
        memory                     = "1Gi"
        env                        = [{
          name                     = "APP_PORT"
          value                    = 3000
        }]
      }]
      min_replicas                 = 1
      max_replicas                 = 1
    }
  },
  {
    name                           = "pythonapp"
    revision_mode                  = "Single"
    dapr                           = {
      app_id                       = "pythonapp"
      app_port                     = 80
    }
    template                       = {
      containers                   = [{
        name                       = "hello-k8s-python"
        image                      = "dapriosamples/hello-k8s-python:latest"
        cpu                        = 0.5
        memory                     = "1Gi"
      }]
      min_replicas                 = 1
      max_replicas                 = 1
    }
  }]
}

resource "azurerm_container_app" "container_app" {
  for_each                     = {for app in var.container_apps: app.name => app}

  name                         = each.key
  resource_group_name          = var.resource_group_name
  container_app_environment_id = azurerm_container_app_environment.managed_environment.id
  tags                         = var.tags
  revision_mode                = each.value.revision_mode

  template {
    dynamic "container" {
      for_each                   = coalesce(each.value.template.containers, [])
      content {
        name                     = container.value.name
        image                    = container.value.image
        args                     = try(container.value.args, null)
        command                  = try(container.value.command, null)
        cpu                      = container.value.cpu
        memory                   = container.value.memory
        
        dynamic "env" {
          for_each               = coalesce(container.value.env, [])
          content {
            name                 = env.value.name
            secret_name          = try(env.value.secret_name, null)
            value                = try(env.value.value, null)
          }
        }
      }
    }
    min_replicas                 = try(each.value.template.min_replicas, null)
    max_replicas                 = try(each.value.template.max_replicas, null)
    revision_suffix              = try(each.value.template.revision_suffix, null)

    dynamic "volume" {
      for_each                   = each.value.template.volume != null ? [each.value.template.volume] : []
      content {
        name                     = volume.value.name
        storage_name             = try(volume.value.storage_name, null)
        storage_type             = try(volume.value.storage_type, null)
      }
    }
  }

 dynamic "ingress" {
    for_each                     = each.value.ingress != null ? [each.value.ingress] : []
    content {
      allow_insecure_connections = try(ingress.value.allow_insecure_connections, null)
      external_enabled           = try(ingress.value.external_enabled, null)
      target_port                = ingress.value.target_port
      transport                  = ingress.value.transport

      dynamic "traffic_weight"  {
        for_each                 = coalesce(ingress.value.traffic_weight, [])
        content {
          label                  = traffic_weight.value.label
          latest_revision        = traffic_weight.value.latest_revision
          revision_suffix        = traffic_weight.value.revision_suffix
          percentage             = traffic_weight.value.percentage
        }
      }
    }
  }

  dynamic "dapr" {
    for_each                     = each.value.dapr != null ? [each.value.dapr] : []
    content {
      app_id                     = dapr.value.app_id
      app_port                   = dapr.value.app_port
      app_protocol               = dapr.value.app_protocol
    }
  }

  dynamic "secret" {
    for_each                     = each.value.secrets != null ? [each.value.secrets] : []
    content {
      name                       = secret.value.name
      value                      = secret.value.value
    }
  }

  lifecycle {
    ignore_changes = [
        tags
    ]
  }
}
```

### Debug Output/Panic Output

```shell
The container app successfully deploys:
module.container_apps.azurerm_container_app.container_app["nodeapp"]: Creation complete after 39s [id=/subscriptions/1a45a694-ae23-4650-9774-89a571c462f6/resourceGroups/AnubisRG/providers/Microsoft.App/containerApps/nodeapp]

If the container app exposes a public or private ingress, the engineer deploying the container app must specify a `traffic_weight` block in the `ingress` block in the `azurerm_container_app` resource even if this not necessary when creating the same resource using Bicep, ARM, Azure CLI, PowerShell, or Azure REST API.
```

### Expected Behaviour

In `azurerm_container_app` the `traffic_weight` block in the `ingress block` should be optional, while now you have to create at least one `traffic_weight` if the `ingress block` is not null.

### Actual Behaviour

If the container app exposes a public or private ingress, the engineer deploying the container app must specify a `traffic_weight` block in the `ingress` block in the `azurerm_container_app` resource even if this not necessary when creating the same resource using Bicep, ARM, Azure CLI, PowerShell, or Azure REST API.

### Steps to Reproduce

Deploy a container app that exposes a private and public ingress, specify a `traffic_weight` block in the `ingress` block in the `azurerm_container_app` resource even if this is not necessary when creating the same resource using Bicep, ARM, Azure CLI, PowerShell, or Azure REST API.

### Important Factoids

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by paolosalvatori**
### Description
This pull request resolves #20537.

### Expected Behaviour

In `azurerm_container_app`, the `traffic_weight` block in the `ingress block` should be optional, while now you have to create at least one `traffic_weight` if the `ingress block` is not null.

### Actual Behaviour

If the container app exposes a public or private ingress, the engineer deploying the container app must specify a `traffic_weight` block in the `ingress` block in the `azurerm_container_app` resource even if this not necessary when creating the same resource using Bicep, ARM, Azure CLI, PowerShell, or Azure REST API.

### New Behaviour

The `traffic_weight` block is now optional.

### Repro steps

I made the change and successfully deployed a container app with an `ingress` block with no `traffic_weight` block in the `azurerm_container_app`.

### CODE REVIEW

**Comment by paolosalvatori** [comment]
cc: @lonegunmanb

**Comment by paolosalvatori** [comment]
I added the `completeWithNoIngressTrafficWeights` test function to the `container_app_resource_test.go` file. Please check and let me know if I need to make other changes. NOTE: I don't think any changes are required in the `expandContainerAppIngressTraffic` and `flattenContainerAppIngressTraffic` as they already return an empty slice when no `traffic_weight` block is specified.

**Review by katbyte — APPROVED** [review]
Thanks @paolosalvatori ! LGTM 🚂

File: internal/services/containerapps/container_app_resource_test.go [hunk 1]
```diff
@@ -472,6 +472,99 @@ resource "azurerm_container_app" "test" {
 `, r.templatePlusExtras(data), data.RandomInteger, revisionSuffix)
 }
 
+func (r ContainerAppResource) completeWithNoIngressTrafficWeights(data acceptance.TestData, revisionSuffix string) string {
```
Comment (katbyte) [inline internal/services/containerapps/container_app_resource_test.go:475]: I missed this: you'll want to add a function using this test config:
```
func TestAccContainerAppResource_completeWithSidecar(t *testing.T) {
	data := acceptance.BuildTestData(t, "azurerm_container_app", "test")
	r := ContainerAppResource{}
	data.ResourceTest(t, r, []acceptance.TestStep{
		{
			Config: r.completeWithSidecar(data, "rev1"),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.ImportStep(),
	})
}
````

Comment (paolosalvatori) [reply internal/services/containerapps/container_app_resource_test.go:475]: Thanks @katbyte I added the `TestAccContainerAppResource_completeWithNoIngressTrafficWeights` function to `container_app_resource_test.go` file.

**Comment by paolosalvatori** [comment]
Hi @katbyte @jackofallops can you please review, approve, and merge this PR? Thanks :) Will this feature be included in the next release? If yes, what is the ETA for the next release? 😄

**Review by katbyte — APPROVED** [review]
Thanks @paolosalvatori - LGTM 🌳

**Pull request merged**